File: src/System.Windows.Forms.Analyzers/cs/tests/Analyzer/MissingPropertySerializationConfiguration/TestData/EdgeCaseScenarios/AnalyzerTestCode.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using System;
using System.ComponentModel;

namespace Test
{
    // Custom IComponent interface in a different namespace
    // This should not be detected by the analyzer
    namespace CustomComponents
    {
        public interface IComponent : IDisposable
        {
            ISite Site { get; set; }
            event EventHandler Disposed;
        }

        public interface ISite : IServiceProvider
        {
            IComponent Component { get; }
            IContainer Container { get; }
            bool DesignMode { get; }
            string Name { get; set; }
        }

        public interface IContainer : IDisposable
        {
            ComponentCollection Components { get; }
            void Add(IComponent component);
            void Add(IComponent component, string name);
            void Remove(IComponent component);
        }

        public class ComponentCollection
        {
            // Implementation omitted
        }

        // Component implementing the custom IComponent
        // Properties here should not be flagged
        public class CustomComponent : CustomComponents.IComponent
        {
            private ISite _site;

            public ISite Site
            {
                get { return _site; }
                set { _site = value; }
            }

            // This should not be flagged because it's from a custom IComponent
            public string CustomProperty { get; set; }

            public event EventHandler Disposed;

            public void Dispose()
            {
                Disposed?.Invoke(this, EventArgs.Empty);
            }
        }
    }

    // Component implementing System.ComponentModel.IComponent
    public class MyComponent : System.ComponentModel.IComponent
    {
        private System.ComponentModel.ISite _site;

        public System.ComponentModel.ISite Site
        {
            get { return _site; }
            set { _site = value; }
        }

        public event EventHandler Disposed;

        // This should not be flagged because it's static
        public static string StaticProperty { get; set; }

        // This should not be flagged because it has a private setter
        public string PrivateSetterProperty { get; private set; }

        // This should not be flagged because it's internal with a private setter
        internal string InternalPrivateSetterProperty { get; private set; }

        // This WOULD be flagged in a normal scenario (public read/write property)
        public string RegularProperty { get; set; }

        public void Dispose()
        {
            Disposed?.Invoke(this, EventArgs.Empty);
        }
    }

    // Base component with properly attributed properties
    public class BaseComponent : System.ComponentModel.IComponent
    {
        private System.ComponentModel.ISite _site;

        public System.ComponentModel.ISite Site
        {
            get { return _site; }
            set { _site = value; }
        }

        public event EventHandler Disposed;

        // Properly attributed with DesignerSerializationVisibility
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public virtual string AttributedProperty { get; set; }

        // Properly attributed with DefaultValue
        [DefaultValue("Default")]
        public virtual string DefaultValueProperty { get; set; }

        public void Dispose()
        {
            Disposed?.Invoke(this, EventArgs.Empty);
        }
    }

    // Derived component with overridden properties
    public class DerivedComponent : BaseComponent
    {
        // These should not be flagged because they are overrides
        // and the base property is already properly attributed
        public override string AttributedProperty { get; set; }
        public override string DefaultValueProperty { get; set; }
    }
}

File: src/System.Windows.Forms.Analyzers/vb/tests/Analyzer/MissingPropertySerializationConfiguration/TestData/EdgeCaseScenarios/AnalyzerTestCode.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
Option Strict On
Option Explicit On

Imports System
Imports System.ComponentModel

Namespace Test

    ' Custom IComponent interface in a different namespace
    ' This should not be detected by the analyzer
    Namespace CustomComponents

        Public Interface IComponent
            Inherits IDisposable

            Property Site As ISite
            Event Disposed As EventHandler
        End Interface

        Public Interface ISite
            Inherits IServiceProvider

            ReadOnly Property Component As IComponent
            ReadOnly Property Container As IContainer
            ReadOnly Property DesignMode As Boolean
            Property Name As String
        End Interface

        Public Interface IContainer
            Inherits IDisposable

            ReadOnly Property Components As ComponentCollection
            Sub Add(component As IComponent)
            Sub Add(component As IComponent, name As String)
            Sub Remove(component As IComponent)
        End Interface

        Public Class ComponentCollection
            ' Implementation omitted
        End Class

        ' Component implementing the custom IComponent
        ' Properties here should not be flagged
        Public Class CustomComponent
            Implements CustomComponents.IComponent

            Private _site As ISite

            Public Property Site As ISite Implements IComponent.Site
                Get
                    Return _site
                End Get
                Set(value As ISite)
                    _site = value
                End Set
            End Property

            ' This should not be flagged because it's from a custom IComponent
            Public Property CustomProperty As String

            Public Event Disposed As EventHandler Implements IComponent.Disposed

            Public Sub Dispose() Implements IDisposable.Dispose
                RaiseEvent Disposed(Me, EventArgs.Empty)
            End Sub
        End Class
    End Namespace

    ' Component implementing System.ComponentModel.IComponent
    Public Class MyComponent
        Implements System.ComponentModel.IComponent

        Private _site As System.ComponentModel.ISite

        Public Property Site As System.ComponentModel.ISite Implements System.ComponentModel.IComponent.Site
            Get
                Return _site
            End Get
            Set(value As System.ComponentModel.ISite)
                _site = value
            End Set
        End Property

        Public Event Disposed As EventHandler Implements System.ComponentModel.IComponent.Disposed

        ' This should not be flagged because it's static
        Public Shared Property StaticProperty As String

        ' This should not be flagged because it has a private setter
        Public Property PrivateSetterProperty As String
            Get
                Return String.Empty
            End Get
            Private Set(value As String)
                ' Do nothing
            End Set
        End Property

        ' This should not be flagged because it's internal with a private setter
        Friend Property InternalPrivateSetterProperty As String
            Get
                Return String.Empty
            End Get
            Private Set(value As String)
                ' Do nothing
            End Set
        End Property

        ' This WOULD be flagged in a normal scenario (public read/write property)
        Public Property RegularProperty As String

        Public Sub Dispose() Implements IDisposable.Dispose
            RaiseEvent Disposed(Me, EventArgs.Empty)
        End Sub
    End Class

    ' Base component with properly attributed properties
    Public Class BaseComponent
        Implements System.ComponentModel.IComponent

        Private _site As System.ComponentModel.ISite

        Public Property Site As System.ComponentModel.ISite Implements System.ComponentModel.IComponent.Site
            Get
                Return _site
            End Get
            Set(value As System.ComponentModel.ISite)
                _site = value
            End Set
        End Property

        Public Event Disposed As EventHandler Implements System.ComponentModel.IComponent.Disposed

        ' Properly attributed with DesignerSerializationVisibility
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Overridable Property AttributedProperty As String

        ' Properly attributed with DefaultValue
        <DefaultValue("Default")>
        Public Overridable Property DefaultValueProperty As String

        Public Sub Dispose() Implements IDisposable.Dispose
            RaiseEvent Disposed(Me, EventArgs.Empty)
        End Sub
    End Class

    ' Derived component with overridden properties
    Public Class DerivedComponent
        Inherits BaseComponent

        ' These should not be flagged because they are overrides
        ' and the base property is already properly attributed
        Public Overrides Property AttributedProperty As String
        Public Overrides Property DefaultValueProperty As String
    End Class
End Namespace
